### ISSUE TITLE
WCF GZIP CustomBinding Errror

### ISSUE THREAD

**Issue by MarkOverstreet** [MarimerLLC/cslaforum#180]
I have a VS.NET 2013 (4.5.2)  app using CSLA.NET that transmits large objects and I am trying to get better performance by using compression which according to the documentation is a new features in WCF 4.5.  So I created the seemingly simple customBindings on both the client and server as seen below.

**Web.config (host remote dataportal)**
`
  <system.serviceModel>

    <services>
        <service name="Csla.Server.Hosts.WcfPortal" behaviorConfiguration="returnFaults">
            <endpoint binding="customBinding" bindingConfiguration="customBinding_IWcfPortal" contract="Csla.Server.Hosts.IWcfPortal"/>
        </service>
    </services>
    
    <bindings>

      <customBinding>
        <binding name="customBinding_IWcfPortal">
          <binaryMessageEncoding compressionFormat="GZip"/>
          <httpTransport authenticationScheme="None"  decompressionEnabled="true" maxReceivedMessageSize="2147483647" maxBufferSize="2147483647"/>
        </binding>
      </customBinding>
      
        </binding>
      </wsHttpBinding>
    </bindings>
    
    
    <behaviors>
      <serviceBehaviors>
        <behavior name="returnFaults">
          <serviceDebug includeExceptionDetailInFaults="true"/>
        </behavior>
      </serviceBehaviors>
    </behaviors>
    
  </system.serviceModel>
`

**app.config (client)**
`
<appSettings>
<add key="CslaDataPortalProxy" value="Csla.DataPortalClient.WcfProxy, Csla"/>
</appSettings>
  <system.serviceModel>
    <client>
      <endpoint name="WcfDataPortal" address="http://localhost:3015/WcfPortal.svc" binding="customBinding" contract="WcfPortal.IWcfPortal"/>
    </client>
    <bindings>
      <customBinding>
        <binding name="customBinding_IWcfPortal">
          <binaryMessageEncoding compressionFormat="GZip" />
          <httpTransport authenticationScheme="None" decompressionEnabled="true" maxReceivedMessageSize="2147483647" maxBufferSize="2147483647"/>
        </binding>
      </customBinding>
    </bindings>
  </system.serviceModel>
`

However, I get the error below when I try and run the solution (if I use basicHttpBinding all works just slooowly).  As you can see, I do have a tag defining httpTransport so I'm wondering if anyone else might have a solution or a suggestion?

**ERROR**
An exception of type 'System.InvalidOperationException' occurred in Csla.dll but was not handled in user code

Additional information: The CustomBinding on the ServiceEndpoint with contract 'IWcfPortal' lacks a TransportBindingElement.  Every binding must have at least one binding element that derives from TransportBindingElement.

Thanks
Mark

**Comment by rockfordlhotka** [comment]
From your other posts I am concerned that you are maybe trying to transfer the image bodies (the actual image) as part of your object's state. I would strongly recommend against that, because with or without CSLA you are likely to run into issues transferring such large blobs via WCF.

It is generally better to set up an http endpoint specifically for the download/upload of images, and then have your object's state include the URL or whatever other information is necessary to interact with that http endpoint.

**Comment by MarkOverstreet** [comment]
You are correct, that is exactly how I have the classes designed right now.  I'm not sure I understand how your suggestion would work.  For example, this is a Winform app that will be executing outside the firewall and currently the objects are serializing over port 80 with no issues other than being really slow.  

So how would I get the image from the user's computer other than trying to get the firewall opened up and using some other transport mechanism like FTP?

**Comment by MarkOverstreet** [comment]
By the way, even if I did change the architecture, I would love to be able to get GZIP working with CSLA.  :)

**Comment by rockfordlhotka** [comment]
Behind the firewall you are hosting the data portal in IIS/ASP.NET, so you have everything you need right there. In that very same appserver project that hosts the data portal you can create an http endpoint (exactly how depends on your version of ASP.NET) that directly download the files based on an http request.

On the client you can use the latest http client implementation (which one depends on your version of .NET) to download the file (into memory if you choose) from that endpoint.

That'll tap into a lower level and generally more efficient http transfer mechanism as compared to WCF. Mostly it'll avoid pretty much _all_ serialization and related overhead, and that's a big deal for blobs!

**Comment by MarkOverstreet** [comment]
Okay I see how that would be more efficient but how do you handle the coordination with the business object.  Let's say I create an .aspx page to upload a file using standard http.  Currently  I have a rule on the business object that only allows a certain role to upload and the file has meta data I need to save to the database, so when the object serializes back to the server and writes to the database, how would the object initiate the upload to the server? How then would I get the uploaded file so as to make that association in the database and how would the end point handling the upload keep in sync with the business rules for authorization?

By the way, I'm using Visual Studio 2013 and .NET 4.5.2.   Thanks again for helping me to understand the proper approach with CSLA.

Mark

**Comment by rockfordlhotka** [comment]
I'd look at creating a command object (probably using ReadOnlyBase) that has a DP_Fetch method marked with `[RunLocal]` to make the call to the remote endpoint to retrieve the image. That command object can be invoked by your business object when it needs the image.

**Comment by MarkOverstreet** [comment]
A little more background and a question...

This is a Winforms app that allows the user to upload images (as well as view them).  So the current program allows them to select an image (or two or three, etc) and add them to a collection of Document objects and then calling Save will send them to the server where the data is stored in a SQL Server database but the image is stored on disk and then the path stored in the database.  

So are you saying that during the save, I run the command object on the client (runlocal) and it merely does something like instantiate a WebClient object and gets the image (e.g through http) or posts the the images?  So I guess I would "post" the image(s) and if successful then follow through with the Save so the object can serialize back to the server and save the data to the database.  If this is correct, my last question would be is there a way to intercept the Save before if calls the appropriate DataPortal method so that I can instantiate the command object and try to transfer the image(s)?

Thanks
Mark

**Comment by rockfordlhotka** [comment]
Yes, you can override Save (or these days SaveAsync) and do/start the upload before saving the related metadata.
